sale_import_base/models/sale_channel_importer.py

```diff
@@ -0,0 +1,296 @@
+            )
+            if not country:
+                raise ValidationError(
+                    _("Missing country {}").format(data["country_code"])
+                )
+            result["country_id"] = country.id
+            if data.get("state_code"):
+                state = self.env["res.country.state"].search(
+                    [("code", "=", data["state_code"]), ("country_id", "=", country.id)]
+                )
+                if not state:
+                    raise ValidationError(
+                        _("Missing State %(state_code)s for country %(country_name)s")
+                        % {
+                            "state_code": data["state_code"],
+                            "country_name": country.name,
+                        }
+                    )
+                result["state_id"] = state.id
+        return result
+
+    def _should_merge_addresses(self, vals_addr_invoice, vals_addr_shipping):
+        return vals_addr_invoice == vals_addr_shipping
+
+    def _process_addresses(
+        self, parent, address_invoice, address_shipping, archive_addresses=True
+    ):
+        vals_addr_invoice = self._prepare_partner(
+            address_invoice, parent.id, archive_addresses
+        )
+        vals_addr_shipping = self._prepare_partner(
+            address_shipping, parent.id, archive_addresses
+        )
+        if self._should_merge_addresses(vals_addr_invoice, vals_addr_shipping):
+            # not technically correct for the shipping addr, but this shouldn't matter
+            vals_addr_invoice["type"] = "invoice"
+            result = self.env["res.partner"].create(vals_addr_invoice)
+            return (result, result)
+        else:
+            vals_addr_invoice["type"] = "invoice"
+            vals_addr_shipping["type"] = "delivery"
+            return (
+                self.env["res.partner"].create(vals_addr_invoice),
+                self.env["res.partner"].create(vals_addr_shipping),
+            )
+
+    def _prepare_sale_line_vals(self, data, sale_order):
+        return [self._prepare_sale_line(line, sale_order) for line in data["lines"]]
+
+    def _prepare_sale_line(self, line_data, sale_order):
+        channel = self.payload_id.sale_channel_id
+        company_id = channel.company_id
+
+        product = self.env["product.product"].search(
+            [
+                ("default_code", "=", line_data["product_code"]),
+                ("product_tmpl_id.company_id", "=", company_id.id),
+            ]
+        )
+        if not product:
+            product = self.env["product.product"].search(
+                [
+                    ("default_code", "=", line_data["product_code"]),
+                    ("product_tmpl_id.company_id", "=", False),
+                ]
+            )
+        if not product:
+            raise ValidationError(
+                _(
+                    "There is no active product with the Internal Reference %(code)s "
```

Comment:
Following the idea the the product search domain can be override, we need to change the "Internal Reference" here by "code".

Reply:
Hum right, I did not think about this.
Well maybe it is best to do as you did then. (We do not want multiple override for such a simple change...)
I'd only rename the `get_product_id` by `get_product` and same for `product_id` variable by `product`
if you don't mind, because I think it is more consistent to the way variable are defined globally in Odoo. (usually with `get_product_id` you would expect an id, and not a recordset) Well, not mandatory, I guess it is also a question of habit/preference, but it seems weird to me !

Reply:
Thinking about it, I think there are pros and cons for both options :
-  the override on the _get_product_domain is easy to write, but we loose this information of the field's name during the raise
- the override on the _get_product has a more explicit raise.. But you have to rewrite the 2 raises in your override

Reply:
And thanks for the tips on the variable names ! :pray: 
I spent my whole life thinking it was mandatory to write records with a "_id" suffix, precisely in order to recognize that the variable is a record/recordset (I don't even remember who told me this or where I read this)

But today I discovered that I was wrong this whole time ! :fearful: 
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#variable-names